회원 가입 기능 추가

hobulian opened the pull request:
회원 가입에 대한 기능을 생성했습니다.

hobulian committed:
회원 가입 기능 추가

f-lab-TJ commented:
1. Commit 메세지 body에 더 상세한 내용을 적어주세요. (어떤 것을 변경했는지, 어떤 것이 추가되었는지, test는 어떻게 수행하였는지.
2. Test code가 없습니다. Unit test를 추가해주세요.

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+        return new ResponseEntity<>(null, null, HttpStatus.CREATED);
+    }
+
+    @GetMapping("/checkDuplicated")
```
boolean을 리턴하는 함수이니 isDuplicated 는 어떨까요?

hobulian replied:
그게 좀 더 직관적일 것 같습니다.

f-lab-TJ commented on src/main/java/com/sns/model/Member.java:
```diff
@@ -0,0 +1,23 @@
+    private String email;
+    private String profile_name;
+    private String profile_message;
+    private String profile_image;
```
profile_imagepath와 차이점이 뭔가요?

hobulian replied:
image는 파일이름이고 imagepath는 파일이 저장되는 경로로 설정했습니다. 지금 생각해보니 파일을 어떻게 어디에 저장할지 부분이 좀 이해가 부족한듯 해서 좀 더 알아보겠습니다.

f-lab-TJ commented on src/main/java/com/sns/model/Member.java:
```diff
@@ -0,0 +1,23 @@
+    private String name;
+    @NonNull
+    private String email;
+    private String profile_name;
```
이름이 없는 경우 빈칸으로 두나요, default 값을 DB에 저장하나요?

hobulian replied:
원래는 빈 칸으로 두기로 했는데, 입력되지 않으면 id와 동일한 값을 저장하는 것도 방법인 듯 합니다.

f-lab-TJ commented on src/main/java/com/sns/model/Member.java:
```diff
@@ -0,0 +1,23 @@
+    private String profile_message;
+    private String profile_image;
+    private String profile_imagepath;
+    private Date joindate;
```
해당 아이템은 서버에서 추가하기로 했던 것 같은데, 로직 추가해봅시다.

hobulian replied:
database에는 추가된 시간으로 들어가게끔 되어있긴 한데, getter 메소드로 가입 시간을 조회하려 할 떄 필요할 듯 해서 변수로 지정해놓았습니다.

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @GetMapping("/checkDuplicated")
+    public ResponseEntity<String> checkDuplicatedMember(@RequestParam String id) {
+        boolean isDuplicated = (memberService.checkDuplicatedMember(id) == 1);
```
유저가 지정한 id를 id로 사용해도 아무런 문제가 없을까요? 유저 id와 별개로 직접 generate한 id를 사용하면 여러가지 장점이 있을 수 있는데, 한번 고민해보면 좋을 것 같습니다.

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @PostMapping("/signup")
+    public ResponseEntity<String> addMember(Member member) {
+        memberService.addMember(member);
```
`checkDuplicated` 후 `signup`하는 것으로 보이는데, 만약 이 두 함수 사이 다른 사람이 해당 id를 선점하면 어떻게 되는 것일까요?

hobulian replied:
해당 사항에 대해서 생각해보지 못했네요. 우선 db는 id에 중복 값이 insert되는 것을 허용하고 있지 않기 때문에 SQLException이 발생할 듯 합니다.

f-lab-TJ commented on src/main/java/com/sns/config/MybatisConfig.java:
```diff
@@ -0,0 +1,49 @@
+import javax.sql.DataSource;
+
+@Configuration
+@Lazy
```
해당 annotation은 왜 사용하였나요?

hobulian replied:
해당 설정들을 lazy initialization하여 애플리케이션 초기 실행 시 성능을 올릴 수 있다고 생각해서 사용했습니다.

hobulian committed:
- Mapper 설정 파일 위치 변경
- 서버측에 유저의 고유 아이디 저장하기 위한 메소드 추가(UUID 활용)
- Member에 ID 속성과 User_ID속성 분리
- DB에 저장된 유저의 실제 ID와 유저가 정한 ID의 중복을 체크하는 SQL 각각 분리
- MyBatisConfig의 @Lazy 제거

hobulian commented:
> 

단위 테스트 작성에 발생한 문제 해결에 시간이 다소 소요되어서 일단 다른 부분들 먼저 수정해서 커밋하였습니다.

hobulian committed:
- application.properties를 application.yml로 변경해서 작성
- Member 클래스에서 image_path제거
- MybatisConfig에서 sqlSessionFactory의 매개변수인 ApplicationContext context를 제거하고 this로 처리
- DB Connect Test코드 작성 및 확인(성공)

hobulian committed:
- 스키마 업로드

f-lab-TJ commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,11 @@
+spring:
```
하나의 pull request에는 하나의 피쳐만 존재해야합니다. 다른 피쳐나 설정파일의 경우 몇줄짜리 수정이더라도 pull request를 따로 만드시기 바랍니다.

f-lab-TJ replied:
여전히 application.yml 파일이 해당 pull request에 존재하네요.

f-lab-TJ commented on bridge_sns_alarm.sql:
```diff
@@ -0,0 +1,58 @@
+CREATE DATABASE  IF NOT EXISTS `bridge_sns` /*!40100 DEFAULT CHARACTER SET utf8mb3 */ /*!80016 DEFAULT ENCRYPTION='N' */;
```
DB 스키마도 다른 pull request로 빼 두도록 합시다.

f-lab-TJ replied:
새로 pull request를 만들면 그곳에 코멘트 작성하도록 하겠습니다.

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,58 @@
+
+        if (isDuplicatedUser) { // User_id(Alias)의 중복 체크
+            return ResponseEntity.badRequest().body("이미 존재하는 ID입니다.");
+        } else {
```
이와 같은 코드 패턴에서 else가 필요할까요?
코드 간결성을 최대한 높이기 위해서 불필요한 else를 줄여 code depth를 줄이도록 합시다.

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    private String generateUniqueId() { // 서버측에 저장할 Unique ID를 DB에서 겹치지 않을 때까지 재귀적으로 생성
```
코멘트와 다르게 재귀적으로 id를 생성하지 않는 것 같습니다.
또, generateUniqueId와 같은 query를 보내는 대신 SQL query를 통해 중복시 insert에 실패하도록 할 수 있는데, 확인해보면 좋을 것 같습니다.

hobulian committed:
- 스키마 업로드 feature를 분리하기 위해서 업로드된 스키마 제거

hobulian committed:
- 스키마 업로드 feature를 분리하기 위해서 업로드된 스키마 제거

hobulian committed:
- 쓸모 없는 코드 제거로 code depth 감소
- UniqueId를 서버에서 제작하지 않고, DB에서 담당하게 하여 서버측의 UUID 제작 기능 제거

hobulian committed:
- Member의 joindate를 created_at으로 변경
- checkDuplicatedID 메소드 삭제
- checkDuplicatedUser의 이름 checkDuplicatedUserID로 변경

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,48 @@
+            return ResponseEntity.badRequest().body("이미 존재하는 ID입니다.");
+        }
+
+        memberService.addMember(member);
```
DB add 가 실패한 경우 에러처리도 해야하지 않을까요?

f-lab-TJ replied:
user id 가 20글자를 넘는 경우는 어떻게 되나요?

hobulian committed:
- MemberController의 Unit Test 작성

hobulian committed:
- Member의 id를 배열로 변경(DB에 Binary로 저장되게 변경한 것에 맞춤)
- MemberController의 Unit Test에 @DisplayName과 주석을 통한 설명 추가

f-lab-TJ commented on src/main/java/com/sns/controller/MemberController.java:
```diff
@@ -0,0 +1,86 @@
+                    if (sqlException.getErrorCode() == 1062) {
+                        isDuplicatedUUID = true;
+                    } else {
+                        throw e;
```
유저에게 노출되는 모든 API는 적절하게 error handling 되어 반환되어야 합니다. 에러를 바로 throw하는 경우, 서버의 디버깅 로그가 사용자에게 전달되는 등의 이슈가 발생할 수 있습니다.

hobulian committed:
- application.yml를 다른 Pull Request로 분리
- 예외를 다루기 위한 Handler 클래스 추가
- DuplicatedUserIdException와 FailedToSignUpException 예외 추가
- org.hibernate.validator:hibernate-validator:7.0.1.Final과 org.hibernate.validator:hibernate-validator-cdi:7.0.1.Final 의존성 추가
- jakarta.validation.constraints를 이용하여, Memeber의 몇몇 요소들에 제약 조건 추가
- MemberController에서 다루던 여러 회원가입 제약 체크 기능들을 MemberServiceImpl로 이전
- 변경된 MemberController에 대해서 기존의 Unit Test 변경

hobulian reviewed (commented):
여러가지 주제가 하나의 커밋에 들어간듯 하여, 커밋을 Revert한 뒤에 주제 별로 나눠서 커밋하였습니다.

hobulian committed:
Revert "- application.yml를 다른 Pull Request로 분리"

This reverts commit 680344ea3da6ddd30970dcf956576c5d1c34bed7.

hobulian committed:
- application.yml를 다른 Pull Request로 분리

hobulian committed:
- 예외를 다루기 위한 Handler 클래스 추가
- DuplicatedUserIdException와 FailedToSignUpException 예외 추가
- org.hibernate.validator:hibernate-validator:7.0.1.Final과 org.hibernate.validator:hibernate-validator-cdi:7.0.1.Final 의존성 추가
- jakarta.validation.constraints를 이용하여, Memeber의 몇몇 요소들에 제약 조건 추가
- MemberController에서 다루던 여러 회원가입 제약 체크 기능들을 MemberServiceImpl로 이전

hobulian committed:
- 변경된 MemberController에 대해서 기존의 Unit Test 변경

f-lab-TJ commented on src/main/java/com/sns/exceptions/Handler.java:
```diff
@@ -0,0 +1,44 @@
+            errors.append(error.getDefaultMessage());
+            errors.append("\n");
+        }
+
```
log4j등을 이용한 로그도 추가해주도록 합시다.

hobulian committed:
- slf4j의 추상체에서 구현된 log4j2를 이용한 Log 추가

f-lab-TJ commented:
Approve를 하였는데 github에서는 Review required로 뜨네요. github 설정을 한번 확인해보고, 머지해서 올려보도록 합시다.

hobulian committed:
- log4j2 설정 파일 추가
